csrc/device_lower/utils.cpp

```diff
@@ -1013,6 +1016,625 @@ bool predicateAtEnd(ForLoop* loop) {
+        }
+      }
+      if (has_unknown) {
+        group_extent = nullptr;
+      }
+      while (count < dq.size()) {
+        dq.pop_front();
+      }
+      // If dq has only one item, we can simplify it to the item itself.
+      if (dq.size() == 1) {
+        group = dq.front();
+      } else {
+        group = std::move(dq);
+      }
+    }
+    // If group has the same extent as linear_g, and there is no inner extent in
+    // group, the the full group represents linear_g.
+    if (new_inner_extent == nullptr && group_extent != nullptr &&
+        simplifyExpr(IrBuilder::eqExpr(group_extent, current.selected_extent))
+            ->isTrue()) {
+      return group;
+    }
+    return PartOf<LinearGroupProjection>{
+        std::make_shared<LinearGroupProjection>(group),
+        new_inner_extent,
+        current.selected_extent};
+  } else {
+    NVF_ERROR(from.size() == 2);
+    NVF_ERROR(to.size() == 1);
+    NVF_ERROR(
+        related(*current.group, from.front()) ||
+        related(*current.group, from.back()));
+    // Adding more extent to the inner.
+    if (current.group->is<ValGroup>() &&
+        current.group->as<ValGroup>() == from.front()) {
+      return PartOf<LinearGroupProjection>{
+          std::make_shared<LinearGroupProjection>(to.front()),
+          SimplifyingIrBuilder::mulExpr(
+              current.inner_extent,
+              from.back()->front()->as<IterDomain>()->extent()),
+          current.selected_extent};
+    }
+    // Adding more extent to the outer.
+    if (current.group->is<ValGroup>() &&
+        current.group->as<ValGroup>() == from.back()) {
+      return PartOf<LinearGroupProjection>{
+          std::make_shared<LinearGroupProjection>(to.front()),
+          current.inner_extent,
+          current.selected_extent};
+    }
+    // Other cases are not implemented yet. Just return std::monostate,
+    // which will make proveLinearAndGetStride stop the propagation and return
+    // "can not prove linear". In the future, we can implement these cases if it
+    // turns out to be useful.
+    return {};
+  }
+}
+
+LinearGroupProjection propagate(
+    const Composition<LinearGroupProjection>& current,
+    const ValGraph& id_graph,
+    const ExprGroup& eg,
+    Direction direction) {
+  auto from = fromGroups(id_graph, eg, direction);
+  auto to = toGroups(id_graph, eg, direction);
+  if (from.size() == 1) {
+    NVF_ERROR(to.size() == 2);
+    NVF_ERROR(std::any_of(current.begin(), current.end(), [&](const auto& g) {
+      return related(g, from.front());
+    }));
+    // Split one group in current into two groups.
+    bool may_be_indivisible_split = eg->front()->isA<Split>() &&
+        !simplifyExpr(eg->front()->as<Split>()->isDivisible())->isTrue();
+    Composition<LinearGroupProjection> result;
+    bool first = true;
+    for (const auto& g : current) {
+      if (g.is<ValGroup>() && g.as<ValGroup>() == from.front()) {
+        if (may_be_indivisible_split && !first) {
```

Comment:
Why `!first`? The first item of the composition correspond to the outermost dimension, right? Why can it be ignored?

Reply:
@naoyam Looks like in order to update this, I need to do a refactor to separate simplification with propagation. I believe this will also make the code easier to read. I will let you know when it is ready to review again.